modules/instance_template/main.tf

```
@@ -42,7 +42,19 @@ locals {
     },
   ]
 
-  all_disks = concat(local.boot_disk, var.additional_disks)
+  all_disks = concat(
+    local.boot_disk, tolist(
+      [
+        for d in var.additional_disks : {
+          disk_size_gb = d.disk_size_gb,
+          disk_type    = d.disk_type,
+          auto_delete  = "true",
+          boot         = "false",
+          type         = replace(replace(d.disk_type, "/pd-*/", ""), "/local-.*/", "SCRATCH")
```

I'm not clear on why we need this. Can you explain more (maybe add a comment)?

Also possible typo
```suggestion
          type         = replace(replace(d.disk_type, "/pd-.*/", ""), "/local-.*/", "SCRATCH")
```

---

sure I'm adding the comment now and thank for eyeing the regex.
The main motive for doing it was to allow reuse of "disk_type" to set "type" which seems to be required in order to use the local SSD. The other disk_type's are all persistent, so we really only need to do funny replace magic to set SCRATCH and use PERSISTENT for everything else.
If this isn't needed I don't mind removing it, I just wanted to add SCRATCH without having other people change their code :-)

---

Got it, I'm still a little confused on how this _fixes_ #76 though. The issue seems to say that this is a bug which can't be worked around, but this fix seems to mostly be for brevity (combining `disk_type` and `type` into one field).

---

Ah, I haven't tried to add other fields like NVME with this change, I could try to do that if you like. I think I had focused more on the subject of #76, which I had largely thought was "instance template with a scratch disk". I had seen the same kind of error message when I tried to add a local-ssd, terraform had complained:
> disk_type of local-ssd must be SCRATCH disks. disk 2 is a  disk
Which, again, lead me here, I don't know why it wasn't seeing that I was including type=SCRATCH... I tried to add debugging to terraform and see what was being set.

Would you like me test this change with other attributes? Or is this PR mostly unneeded?
It really isn't my intention to add unnecessary complication, but I couldn't get local disks to be created without doing this

---

I'm fine with the change, but I'm mostly just trying to understand why it is necessary. As far as I can tell, we should already be surfacing the disk type info to Terraform.

Do you mind sharing your config which *wasn't* working before but now is working with this PR?

---

sure, I'll put a snippet here, and a more complete (should be runnable, I had to remove some items) here: https://gist.github.com/kondor6c/b6905e76fbdc2c70dcb0d95d2c43d9a3
I'm encapsulating this (terraform-google-vm) in another module, which isn't normally something I like doing, but this module does a lot of what I'm trying to do :-)


locals.tf:
```
local_ssd_types = [ "n2", "n2d", "n1", "c2", "m2", "a2" ]
host_disk = tolist([
    for d in split("-", var.scaling_class) : 
      { 
        disk_size_gb = "375",
        boot = "false",
        auto_delete = "false",
        disk_type = "local-ssd",
        type = "SCRATCH"
      } if contains(local.local_ssd_types, d) ])
disks = tolist([
  for i in var.extra_disks: {
    disk_size_gb = i.extra_disk_size,
    disk_type = i.extra_disk_type,
    auto_delete = "true",
    boot = "false",
    type = "PERSISTENT"
   }
 ])
```
main.tf:
```
tags                 = concat([var.project_name],  var.tags)
metadata             = merge(var.metadata, {"VmDnsSetting" = "ZonalPreferred", "user-data" = data.template_cloudinit_config.cinit.rendered } )
startup_script       = data.template_file.provider_bootstrap_script.rendered
auto_delete          = true
additional_disks     = concat(local.disks, local.host_disk)
```

The _same_ exact config works once the change is made. I used terraform console to ensure the crazy for loops were being created properly. If you prefer to communicate using a different method, I have linked my keybase ID (which holds other identities) to my github ID.

---

What is the terraform plan output for your config?

It's just very weird to me because it looks like `type = "PERSISTENT"` should be properly set in your config so I don't know what bug you'd be encountering it. Dynamically generating the same value you're already setting manually shouldn't be a fix.

---

If you would like to keep this change, we should make sure the computed value doesn't *replace* a `type` that was explicitly passed.